### PR TITLE
Implement net files support for NNUE engines

### PR DESCRIPTION
**Pull request by vshcherbyna**
This PR implements support for network files (for NNUE based engines) in OpenBench:

- Test model object is modified and it contains two new fields: devnetwork and basenetwork (256 characters max)
- Client script is modified to download devnetwork and basenetwork into 'Engines' directory and pass as option.EvalFile to an engine
- In case Engines folder already contains eval file the download is skipped to preserve bandwidth
- UI is modified to allow net URLs to be passed during creation of a test
- UI is modified to view net URLs to per each test

This is how UI looks like now

Creation of new test:

image

Seeing existing test without devnetwork and basenetwork parameters:

image

Seeing existing test with devnetwork and basenetwork parameters:

image

This PR is tested on my local open bench instance http://shcherbyna.com:8000/index/ using Windows 10 and CentOS/Ubuntu/Debian clients.

Important: the changes on back end are not going to break old clients because they will ignore basenetwork and devnetwork fields and will continue working.

### CODE REVIEW

File: Client/Client.py [hunk 1]
```diff
@@ -284,12 +299,20 @@ def getCutechessCommand(arguments, data, nps):
         '{0}-{1}'.format(data['test']['engine'], data['test']['dev']['name'])
     )
 
+    # In case devnetwork is specified append EvalFile option
```
Comment (AndyGrant) [inline Client/Client.py:302]: Change the comment to

``# Add evaluation weights for the Dev engine if specified ``

File: Client/Client.py [hunk 2]
```diff
@@ -284,12 +299,20 @@ def getCutechessCommand(arguments, data, nps):
...
+    if devnetwork:
+        devflags = devflags + " option.EvalFile=" + devnetwork
+
     # Options for the Base Engine
     baseflags = '-engine dir=Engines/ cmd=./{0} proto={1} tc={2}{3} name={4}'.format(
         baseCommand, data['test']['base']['protocol'], timecontrol, baseoptions,
         '{0}-{1}'.format(data['test']['engine'], data['test']['base']['name'])
     )
 
+    # In case basenetwork is specified append EvalFile option
```
Comment (AndyGrant) [inline Client/Client.py:312]: Change the comment to

``# Add evaluation weights for the Base engine if specified ``

File: Client/Client.py [hunk 3]
```diff
@@ -257,6 +268,10 @@ def getCutechessCommand(arguments, data, nps):
     devCommand = addExtension(data['test']['dev']['sha'])
     baseCommand = addExtension(data['test']['base']['sha'])
 
+    # Download network file(s) if configured
```
Comment (AndyGrant) [inline Client/Client.py:271]: I'de like to see this as:

```
    devnetwork  = getNetworkWeights(data['test']['dev']['network'])
    basenetwork = getNetworkWeights(data['test']['base']['network'])
```

File: Client/Client.py [hunk 4]
```diff
@@ -241,6 +240,18 @@ def getEngine(data, engine):
...
+
+    if targetnet:
+        localnet = pathjoin('Engines', os.path.basename(targetnet)).rstrip('/')
+        # Check if a given network file is present in Engines directory
+        if os.path.isfile(localnet):
+            return os.path.basename(targetnet)
+        # Download network file
+        getFile(targetnet, localnet)
+        return os.path.basename(targetnet)
+
```
Comment (AndyGrant) [inline Client/Client.py:254]: I'm very picky ....

```
def getNetworkWeights(network):

    if network:
        fname = pathjoin('Engines', os.path.basename(network)).rstrip('/')
        if not os.path.isfile(fname): getFile(network, fname)
        return os.path.basename(network)

    return None # Comment for you, not the Repo: I know this is implicit, but I want to see None anyway.
```

**Review by AndyGrant — CHANGES_REQUESTED** [review]

**Comment by AndyGrant** [comment]
In https://github.com/AndyGrant/OpenBench/blob/master/OpenBench/utils.py#L176, you can add a new function verifyGithubLink, and then add two lines to the verification array like this: ``(verifyGithubLink, 'devnetwork', 'Dev Network'),``

```
def verifyGithubLink(field, fieldName):
    if request.POST[field] and not request.post[field].startswith('https://github.com/'):
        errors.append('{0} must be located on a Github URL'.format(fieldName)
```

If you add a new test with fields NOT on github, it should reject it, return you to the homepage, and present a list of errors.
Something like 

```
[0] Dev Network must be located on a Github URL
[1] Base Network must be located on a Github URL
```

Make sure that an empty field is allowed.

File: OpenBench/utils.py [hunk 5]
```diff
@@ -257,6 +257,8 @@ def createNewTest(request):
     test.upperllr    = math.log((1.0 - test.beta) / test.alpha)
     test.dev         = getEngine(*devinfo, protocol)
     test.base        = getEngine(*baseinfo, protocol)
+    test.devnetwork  = request.POST['devnetwork']
```
[the remark this answers is not in the document]
Comment (vshcherbyna) [reply OpenBench/utils.py]: Fixed.

File: OpenBench/utils.py [hunk 6]
```diff
@@ -390,6 +391,8 @@ def workloadDictionary(test, result, machine):
                 'sha'       : test.base.sha,    'bench'     : test.base.bench,
                 'options'   : test.baseoptions,
             },
+            'devnetwork'    : test.devnetwork,
```
[the remark this answers is not in the document]
Comment (vshcherbyna) [reply OpenBench/utils.py]: Fixed.

File: Client/Client.py (hunk 1 above)
Comment (vshcherbyna) [reply Client/Client.py]: Fixed.

File: Client/Client.py (hunk 2 above)
Comment (vshcherbyna) [reply Client/Client.py]: Fixed.

File: Client/Client.py (hunk 3 above)
Comment (vshcherbyna) [reply Client/Client.py:268]: Fixed.

File: Client/Client.py (hunk 4 above)
Comment (vshcherbyna) [reply Client/Client.py]: Fixed.

**Comment by vshcherbyna** [comment]
> 
> 
> In https://github.com/AndyGrant/OpenBench/blob/master/OpenBench/utils.py#L176, you can add a new function verifyGithubLink, and then add two lines to the verification array like this: `(verifyGithubLink, 'devnetwork', 'Dev Network'),`
> 
> ```
> def verifyGithubLink(field, fieldName):
>     if request.POST[field] and not request.post[field].startswith('https://github.com/'):
>         errors.append('{0} must be located on a Github URL'.format(fieldName)
> ```
> 
> If you add a new test with fields NOT on github, it should reject it, return you to the homepage, and present a list of errors.
> Something like
> 
> ```
> [0] Dev Network must be located on a Github URL
> [1] Base Network must be located on a Github URL
> ```
> 
> Make sure that an empty field is allowed.

Thanks for the suggestion and the code snippet, I've implemented this and tested three main scenarios:

1. No dev net or base net present - test creation works
2. There is dev net or base net but it is not on github https page - test creation fails with a correct error
3. There is a dev net or base net and it is on the github page - test creation works

**Pull request merged**